Added electric pair mode, also added braces as a custom pair

justinjk007 opened the pull request:
Turn on electric pair mode by default, also added braces as a custom pair which automatically hooks to major modes like C, C++= and so on. 

We can also define other characters *, / for org-mode if you want, but I see org-mode isn't on the list(yet?).

justinjk007 committed:
Added electric pair mode, also added braces as a custom pair

josteink commented:
I don't know if I'm the ex exception, but I don't like having electric pair set globally. 

How about having it constrained to prog-mode instead?

angrybacon commented:
@josteink I think it's still useful for quotes and parentheses inside `text-mode`. When writing Markdown content too, brackets are often used. The minibuffer also benefits from it.

ssunday commented:
Personally, I don't like electric mode on at all–I find it more trouble than it's worth and many people I know agree. Just throwing in my 2cents

justinjk007 commented:
Well, ok so if you want that constrained to prog-mode ..sure, but all these web developers who want to write css, css is not in prog-mode and like @angrybacon mentioned marked down will also benefit from them.

justinjk007 commented:
@ssunday so just type both of the brackets instead of one ? interesting... That is 2 against 2 votes now. May be this is why emacs devs left it, but then again coming from other popular text editors this is ON by default.

ssunday commented:
@justinjk007 Yes, I like manually typing out the brackets. I tried to use the pair mode but it always did what I didn't want to happen. In my experience, it was problematic refactoring Clojure code.

justinjk007 commented:
@ssunday sometimes when editing lisp this is also a problem .I can see where you're​ coming from, adding and removing parens in the middle is also problematic for me sometimes but it wasn't so much bothering. Hmm may be that's why the devs left it untouched I guess, who are more custom to editing elisp.

justinjk007 commented:
Well... I guess we don't need this then..

angrybacon commented:
People who work often with Lisp generally use a smarter package, there are many: Smartparens, Paredit, Lispy. Of the top of my head, I don't actually know what these packages do, except for Smartparens. But I'm not so heavy on Lisp.

justinjk007 committed:
Added show-paren-mode to defaults

justinjk007 committed:
Revert "Added electric pair mode, also added braces as a custom pair"

This reverts commit 363072216c147042a2afe888583178d805202ec6.

justinjk007 commented:
@josteink Ok I have removed electric pairs and added show parens mode by default can you merge this now to the basic-configurations branch. Thank you.

josteink commented:
>  css is not in prog-mode

The same applies to XML and a few other modes. My solution to that has to manually my add my own `prog-mode-hook` to these modes too. I think that is a far better solution than setting programming-settings globally.

A even better solution would probably to fix this at the root and get the Emacs core developers to admit CSS and XML is more related to programming than text, but altering that now  might be considered a "breaking change"?

As for your final settings... I think that should be fairly uncontroversial. Feel free to merge it yourself :)
